File: src/tables/wide/block.ts
// src/tables/wide/block.ts
//
// The wide parser's JS side: instantiate, hand it a block of whole rows, lift
// the active planes out of the slab. Kept apart from worker.ts because this
// is the riskiest index arithmetic in the build (a row stride the MODULE
// reports), so tests/test_ingest_interface.mjs drives it directly in Node.

import { KEY_COLS } from './header';

/** block.c's ABI_VERSION. */
export const PARSER_ABI = 3;

/**
 * What a parser build is fixed at: two byte budgets, read from the module.
 * Nothing about width: blocks are cut in bytes, so rows-per-block and width
 * are inversely related and the slab needs about the same room at any width.
 * Only the ABI says what the slab MEANS, so only the ABI is checked strictly.
 */
export interface ParserBudget {
  /** block.c's `ARENA_BYTES`: the scratch one block's output is laid out in. */
  arenaBytes: number;
  /** block.c's `BLOCK_BYTES`: the input window one block is copied into. */
  inbufBytes: number;
  /** block.c's `ABI_VERSION`, already checked against PARSER_ABI. */
  abi: number;
}

/** The shape one block's slab is configured to: a runtime split of the arena,
 * exactly as wide as the file's header. */
export interface SlabLayout {
  metrics: number;
  /** Rows one block may carry at that width. */
  rows: number;
}

/** Arena bytes per row at `metrics` planes (floats + u16 hour + u8 TOU);
 * mirrors configure() in block.c. */
function rowCost(metrics: number): number {
  return metrics * 4 + 3;
}

/** Rows the arena holds at `metrics` planes, or 0. Blocks are sized from this,
 * never a constant, so a different arena needs no other change. */
export function maxRowsAt(budget: ParserBudget, metrics: number): number {
  if (!(metrics > 0)) return 0;
  return Math.floor(budget.arenaBytes / rowCost(metrics));
}

/** An instantiated parser and the budgets it reported. */
export interface Parser {
  exports: ParserExports;
  budget: ParserBudget;
}

/**
 * Marker on the slab overflow error. Workers post only `error.message`, so
 * this string is how the retry recognises the one failure it can fix. Never
 * localise or reword it.
 */
export const OVERFLOW_MARKER = 'block-slab-overflow';

/** True for the overflow error above, wherever it has been stringified. */
export function isOverflowError(message: string): boolean {
  return message.includes(OVERFLOW_MARKER);
}

import { afterNextNewline, NEWLINE } from '../../ingest';

// Re-exported for worker.ts and the ingest tests.
export { afterNextNewline, NEWLINE };

export interface ParserExports {
  memory: WebAssembly.Memory;
  inbuf_ptr(): number;
  inbuf_size(): number;
  slab_ptr(): number;
  row_hour_ptr(): number;
  row_tou_ptr(): number;
  arena_bytes(): number;
  /** The layout currently configured; 0 before the first `configure`. */
  slab_rows(): number;
  slab_metrics(): number;
  abi_version(): number;
  last_rows(): number;
  last_overflow(): number;
  last_wide_field(): number;
  last_bad_row(): number;
  last_feb29(): number;
  last_year_mismatch(): number;
  /** Lay the arena out as `numMetrics x maxRows`. Returns 0 if it will not
   * fit, and leaves the regions null so a parse writes nothing. */
  configure(numMetrics: number, maxRows: number): number;
  slab_fill_nan(rows: number): void;
  parse_block(len: number, year: number): number;
}

/** One parsed block as a ROW LIST: row `r` holds the values at `rowHour[r]`,
 * in byte order, which means nothing. */
export interface BlockPayload {
  /** Valid rows emitted; Feb 29 and refused rows are counted apart. */
  rows: number;
  /** data[p * rows + r], plane-major so one entity is a contiguous copy. */
  data: Float32Array;
  /** Hour of the year each row lands on, 0..8759. Length `rows`. */
  rowHour: Uint16Array;
  /** Per-row TOU code from the file. Length `rows`. */
  rowTou: Uint8Array;
  /** Feb 29 rows dropped (intended, and stated). */
  feb29: number;
}

/** A block that contributed nothing, freshly allocated every time (see
 * src/tables/long/block.ts). */
function emptyPayload(feb29 = 0): BlockPayload {
  return {
    rows: 0,
    data: new Float32Array(0),
    rowHour: new Uint16Array(0),
    rowTou: new Uint8Array(0),
    feb29,
  };
}

/**
 * Instantiate the parser and read its budgets; every worker needs its own
 * instance (no SharedArrayBuffer on GitHub Pages). The ABI is CHECKED, the
 * budgets READ: any arena can be driven within its budget, but a slab that
 * MEANS something else cannot be driven correctly at any size.
 */
export async function instantiateParser(module: WebAssembly.Module): Promise<Parser> {
  const instance = await WebAssembly.instantiate(module, {});
  const exports = instance.exports as unknown as ParserExports;
  // A stale block.wasm must fail HERE, before a block is parsed: an older
  // slab layout gives wrong numbers, not an error.
  if (typeof exports.abi_version !== 'function' || exports.abi_version() !== PARSER_ABI) {
    const found = typeof exports.abi_version === 'function' ? exports.abi_version() : 'none';
    throw new Error(
      `block.wasm reports ABI ${found}, but this build speaks ABI ${PARSER_ABI}. ` +
        `Rebuild the parser: parser/wide/build.sh.`,
    );
  }
  const budget: ParserBudget = {
    arenaBytes: exports.arena_bytes(),
    inbufBytes: exports.inbuf_size(),
    abi: exports.abi_version(),
  };
  // A zero budget would surface much later as "file too wide"; refuse now.
  if (!(budget.arenaBytes > 0) || !(budget.inbufBytes > 0)) {
    throw new Error(
      `block.wasm reports arena_bytes()=${budget.arenaBytes}, ` +
        `inbuf_size()=${budget.inbufBytes}. Both must be positive; this binary ` +
        `cannot be driven.`,
    );
  }
  return { exports, budget };
}

/**
 * Parse `bytes[from, to)` (a row boundary to just after a `\n`) and lift
 * `activePlanes`. Every row carries its own hour, so blocks and rows may come
 * in any order. Rows whose year differs from `year` are refused.
 */
export function parseBytes(
  parser: Parser,
  layout: SlabLayout,
  bytes: Uint8Array,
  from: number,
  to: number,
  activePlanes: Int32Array,
  year: number,
): BlockPayload {
  const length = to - from;
  if (length <= 0) return emptyPayload();

  const { exports } = parser;
  const { metrics: slabMetrics, rows: slabRows } = layout;

  // Configured in the same call that parses, so every stride below is the
  // shape just asked for and no block can inherit a stale one. It is cheap.
  if (!exports.configure(slabMetrics, slabRows)) {
    throw new Error(
      `The parser's ${parser.budget.arenaBytes} B arena cannot hold a ` +
        `${slabMetrics} x ${slabRows} slab (${slabMetrics * slabRows * 4 + slabRows * 3} B ` +
        `needed). Cut smaller blocks or build the parser with a larger ARENA_MIB.`,
    );
  }

  const inbufPtr = exports.inbuf_ptr();
  const inbufSize = exports.inbuf_size();
  if (length + 1 > inbufSize) {
    throw new Error(`Block of ${length} B exceeds the ${inbufSize} B WASM input window.`);
  }

  const memory = new Uint8Array(exports.memory.buffer);
  memory.set(bytes.subarray(from, to), inbufPtr);
  let padded = length;
  // parse_block emits a row at its terminator; add one so the last row counts.
  if (memory[inbufPtr + padded - 1] !== NEWLINE) memory[inbufPtr + padded++] = NEWLINE;

  // Clear the whole configured slab: a stale float from the previous block
  // would be a plausible wrong number.
  exports.slab_fill_nan(slabRows);
  const rows = exports.parse_block(padded, year);

  // Counters are read before the empty-block shortcut: a block whose every
  // row was refused has the most to report.
  const wide = exports.last_wide_field();
  if (wide > 0) {
    throw new Error(
      `${wide} field(s) sit past column ${slabMetrics + KEY_COLS}, but this file's header ` +
        `declared ${slabMetrics} entity columns. Rows carrying more fields than the ` +
        `header would be dropped silently, so the load is refused instead.`,
    );
  }
  const bad = exports.last_bad_row();
  if (bad > 0) {
    throw new Error(
      `${bad} row(s) carry a Date or Hour this parser could not read (expected M/D/YYYY and ` +
        `an hour-ending 1-24). Those rows would be dropped silently, so the load is refused.`,
    );
  }
  const mismatched = exports.last_year_mismatch();
  if (mismatched > 0) {
    throw new Error(
      `${mismatched} row(s) carry a year other than ${year}, which is the year on this file's ` +
        `first data row. A case is one calendar year of 8,760 hours, so a second year would be ` +
        `folded onto the same hours. Split the export by year and load the files separately.`,
    );
  }
  const overflow = exports.last_overflow();
  if (overflow > 0) {
    throw new Error(
      `${OVERFLOW_MARKER}: ${overflow} row(s) past the ${slabRows}-row block slab — this ` +
        `block held more rows than its byte length predicted.`,
    );
  }

  const feb29 = exports.last_feb29();
  if (rows === 0) return emptyPayload(feb29);

  // Plane-major on both sides, so each entity is one contiguous copy.
  const slab = new Float32Array(exports.memory.buffer, exports.slab_ptr(), slabMetrics * slabRows);
  const data = new Float32Array(activePlanes.length * rows);
  for (let p = 0; p < activePlanes.length; p++) {
    const src = activePlanes[p] * slabRows;
    data.set(slab.subarray(src, src + rows), p * rows);
  }

  const rowHour = new Uint16Array(exports.memory.buffer, exports.row_hour_ptr(), slabRows);
  const rowTou = new Uint8Array(exports.memory.buffer, exports.row_tou_ptr(), slabRows);

  return {
    rows,
    data,
    // Copied, not viewed: transferred to the main thread, and the next block
    // reuses this memory.
    rowHour: rowHour.slice(0, rows),
    rowTou: rowTou.slice(0, rows),
    feb29,
  };
}

File: parser/wide/block.c
// parser/wide/block.c
//
// The WASM CSV parser for WIDE-SHAPE exports (shape W). A SHAPE reader: area,
// interface, bus and generator exports all come in this shape and nothing
// here can tell them apart. Every ABI argument is a NUMBER about bytes and
// offsets, so `if (kind == BUS)` is unrepresentable (tests/test_wide_abi.mjs).
//
// Input shape; JS consumes the preamble and header, so this sees data rows only:
//
//   line 1   <Entity> Hourly 'Power Flow (MW)' Data for Year 2034
//   line 2-4 preamble (bus exports add an id row, pushing the header to 6)
//   line 5   Date, Hour, TOU,<entity 1>,<entity 2>,...
//   line 6+  one row per HOUR
//
// Invariants:
//
// 1. MEMORY is a function of BLOCK size, never case size (every Worker has
//    its own instance): a fixed input window plus a fixed arena, split between
//    width and rows at runtime.
// 2. BLOCK INDEPENDENCE. Each row's hour comes from its own Date and Hour,
//    never a row counter, so blocks parse in any order.
// 3. ONE ROW = ONE HOUR, so no hour spans two blocks.
// 4. ROW ORDER DOES NOT MATTER. Output is a row list with each row's hour in
//    rowHour[r]; the main thread scatters it.
//
// Build: ./build.sh

#include <wasm_simd128.h>

// ---------------------------------------------------------------- dimensions
//
// Two byte budgets are the only build parameters (`-D` from build.sh):
// `BLOCK_BYTES` (input window) and `ARENA_BYTES` (one block's output). The
// defaults are the shipping build that block.wasm is committed from.
//
// The slab's SHAPE is not compiled in: `configure(numMetrics, maxRows)`
// re-points the output regions inside the arena and refuses a layout that
// does not fit. Width and rows-per-block are inversely related (blocks are cut
// in bytes), so one budget serves every width: an 8 MiB block needs about the
// same slab at 215 entities as at 5,900. Sizing both as independent maxima
// would make bus width unaffordable.
//
// block.ts configures in the same call that parses, so it always indexes with
// the shape it asked for: a wrong stride leaves every number plausible.

// The input window one block is copied into.
#ifndef BLOCK_BYTES
#define BLOCK_BYTES     (12u * 1024u * 1024u)
#endif

// One block's values, row hours and row TOU codes: about 2.7x what an 8 MiB
// block needs at any width. A layout that does not fit makes configure()
// fail, which JS turns into an error rather than a truncated parse.
#ifndef ARENA_BYTES
#define ARENA_BYTES     (9u * 1024u * 1024u)
#endif

#define KEY_COLS        3           // Date, Hour, TOU

// A bad -D fails at compile time, not as a wrong number in a browser.
_Static_assert(BLOCK_BYTES >= 1024u * 1024u,
               "BLOCK_BYTES below 1 MiB cannot hold a block the JS side dispatches");
_Static_assert(ARENA_BYTES >= 1024u * 1024u,
               "ARENA_BYTES below 1 MiB cannot hold one block's values");

// Bumped on any change to the exported surface or the slab's meaning, so a
// stale committed binary fails at instantiate rather than misparsing.
#define ABI_VERSION     3u

static unsigned char inbuf[BLOCK_BYTES];
static unsigned char arena[ARENA_BYTES] __attribute__((aligned(16)));

// Arena regions, re-pointed by configure(); null until it succeeds.
// slab[m * g_maxRows + r] is PLANE-major, so one entity's block of values is a
// contiguous copy on the JS side.
static float*          slab;
// Where each emitted row lands: hour of the year, 0..8759.
static unsigned short* rowHour;
// Per-row TOU code, 0 = OffPeak, 1 = OnPeak. TOU is FILE DATA, never
// recomputed from the calendar (the real rule varies by tariff and holidays);
// read here only so the bytes are scanned once.
static unsigned char*  rowTou;

// The configured layout. Both 0 until configure() succeeds, which makes an
// unconfigured parse_block a no-op rather than a write through null.
static unsigned g_metrics, g_maxRows;

static unsigned g_rows, g_overflow, g_wideField, g_badRow, g_feb29, g_yearMismatch;

__attribute__((export_name("inbuf_ptr")))     unsigned char* inbuf_ptr(void)    { return inbuf; }
__attribute__((export_name("inbuf_size")))    unsigned       inbuf_size(void)   { return BLOCK_BYTES; }
__attribute__((export_name("arena_bytes")))   unsigned       arena_bytes(void)  { return ARENA_BYTES; }
__attribute__((export_name("slab_ptr")))      float*         slab_ptr(void)     { return slab; }
__attribute__((export_name("row_hour_ptr")))  unsigned short* row_hour_ptr(void){ return rowHour; }
__attribute__((export_name("row_tou_ptr")))   unsigned char* row_tou_ptr(void)  { return rowTou; }
// The layout currently configured (0 before the first configure()).
__attribute__((export_name("slab_rows")))     unsigned       slab_rows(void)    { return g_maxRows; }
__attribute__((export_name("slab_metrics")))  unsigned       slab_metrics(void) { return g_metrics; }
__attribute__((export_name("abi_version")))   unsigned       abi_version(void)  { return ABI_VERSION; }
__attribute__((export_name("last_rows")))     unsigned       last_rows(void)    { return g_rows; }
// Rows past the configured maxRows. JS sizes blocks so this cannot happen and
// shrinks them if it does; a backstop, not a normal outcome.
__attribute__((export_name("last_overflow")))   unsigned     last_overflow(void){ return g_overflow; }
// Fields past the configured width, and rows whose Date/Hour did not resolve.
// Both would be silent data loss if unreported. Feb 29 is counted apart,
// because dropping it is intended.
__attribute__((export_name("last_wide_field"))) unsigned     last_wide_field(void){ return g_wideField; }
__attribute__((export_name("last_bad_row")))    unsigned     last_bad_row(void)  { return g_badRow; }
__attribute__((export_name("last_feb29")))      unsigned     last_feb29(void)    { return g_feb29; }
// Rows whose year differs from the file's first data row. date_to_day ignores
// the year, so without this a two-year file would fold onto the same 8,760
// hours. Counted here, refused by JS.
__attribute__((export_name("last_year_mismatch"))) unsigned  last_year_mismatch(void){ return g_yearMismatch; }

// Cumulative days before each month, non-leap; Feb 29 rows are skipped.
static const unsigned short CUM[12] = {0,31,59,90,120,151,181,212,243,273,304,334};

// f64 digit accumulation. Do NOT swap in fast_float's Eisel-Lemire: it lost on
// wasm32 when measured. Re-measure before believing either result.
// Exponent notation is real (`2.568664E-03` for near-zero values) and must
// parse: NaN would silently read as an absent cell.
static inline float parse_float(const unsigned char* p, const unsigned char* e) {
  if (e <= p) return 0.0f / 0.0f;
  int neg = 0;
  if (*p == '-') { neg = 1; p++; }
  else if (*p == '+') { p++; }

  // Integer part, then `frac / scale`. This rounds twice where strtod rounds
  // once: a 1-ulp float32 difference on a few cells of the widest columns.
  // MEASURED, do not "fix": one f64 mantissa with a single division costs 20%
  // throughput and changes none of those cells (they carry 18 significant
  // digits). The reference comparison gates this at <= 1 ulp.
  double ip = 0.0;
  while (p < e) {
    unsigned d = (unsigned)(*p - '0');
    if (d > 9) break;
    ip = ip * 10.0 + (double)d;
    p++;
  }
  double v = ip;
  if (p < e && *p == '.') {
    p++;
    double f = 0.0, sc = 1.0;
    while (p < e) {
      unsigned d = (unsigned)(*p - '0');
      if (d > 9) break;
      f = f * 10.0 + (double)d;
      sc *= 10.0;
      p++;
    }
    v += f / sc;
  }
  if (p < e && (*p == 'e' || *p == 'E')) {
    p++;
    int eneg = 0;
    if (p < e && (*p == '-' || *p == '+')) { eneg = (*p == '-'); p++; }
    int ex = 0;
    while (p < e) {
      unsigned d = (unsigned)(*p - '0');
      if (d > 9) break;
      ex = ex * 10 + (int)d;
      if (ex > 400) ex = 400;   // past f32 range either way
      p++;
    }
    // Exponentiation by squaring: no libm in a freestanding module.
    double scale = 1.0, base = 10.0;
    for (int k = ex; k; k >>= 1) { if (k & 1) scale *= base; base *= base; }
    v = eneg ? v / scale : v * scale;
  }

  // Leftover characters make the field NaN: a partial parse of a mangled
  // field would be a plausible wrong number.
  if (p != e) return 0.0f / 0.0f;
  return (float)(neg ? -v : v);
}

static inline unsigned parse_uint(const unsigned char* p, const unsigned char* e) {
  unsigned v = 0;
  for (; p < e; p++) {
    unsigned d = (unsigned)(*p - '0');
    if (d > 9) break;
    v = v * 10u + d;
  }
  return v;
}

// M/D/YYYY -> day-of-year, with the year returned for the same-year check.
// FEB29 (dropped on purpose) and NO_DAY (unreadable) are distinct, so the
// leap day never looks like a mangled file.
#define FEB29  0xFFFFFFFEu
#define NO_DAY 0xFFFFFFFFu
static inline unsigned date_to_day(const unsigned char* p, const unsigned char* e,
                                   unsigned* yearOut) {
  unsigned month = 0, day = 0, year = 0;
  while (p < e && *p != '/') { month = month * 10u + (unsigned)(*p - '0'); p++; }
  p++;
  while (p < e && *p != '/') { day = day * 10u + (unsigned)(*p - '0'); p++; }
  p++;
  while (p < e) {
    unsigned d = (unsigned)(*p - '0');
    if (d > 9) break;
    year = year * 10u + d;
    p++;
  }
  *yearOut = year;
  if (month < 1 || month > 12 || day < 1 || day > 31) return NO_DAY;
  if (month == 2 && day == 29) return FEB29;   // Feb 29 dropped
  return CUM[month - 1] + day - 1;
}

/**
 * Lay the arena out for one block: `numMetrics` planes of `maxRows` rows, both
 * taken from the file. Returns 0 if it does not fit, which the caller MUST
 * treat as an error; the regions are then left null so nothing is written at
 * a stale stride.
 */
__attribute__((export_name("configure")))
unsigned configure(unsigned numMetrics, unsigned maxRows) {
  g_metrics = 0u; g_maxRows = 0u;
  slab = 0; rowHour = 0; rowTou = 0;
  if (numMetrics == 0u || maxRows == 0u) return 0u;

  // 64-bit: metrics * rows overflows 32 bits before the arena check could
  // catch it.
  unsigned long long cells = (unsigned long long)numMetrics * maxRows;
  unsigned long long need = cells * 4ull + (unsigned long long)maxRows * 3ull;
  if (need > (unsigned long long)ARENA_BYTES) return 0u;

  // The slab is whole floats, so rowHour starts 4-aligned (it needs 2).
  unsigned off = 0u;
  slab    = (float*)(arena + off);           off += (unsigned)cells * 4u;
  rowHour = (unsigned short*)(arena + off);  off += maxRows * 2u;
  rowTou  = arena + off;

  g_metrics = numMetrics;
  g_maxRows = maxRows;
  return 1u;
}

/**
 * NaN-fill the value planes for `rows` rows. A row with fewer fields than the
 * header leaves later planes untouched, and a stale float from the previous
 * block would be a plausible wrong number.
 */
__attribute__((export_name("slab_fill_nan")))
void slab_fill_nan(unsigned rows) {
  const unsigned maxRows = g_maxRows;
  if (maxRows == 0u) return;                 // never configured; nothing to fill
  if (rows > maxRows) rows = maxRows;
  const float nan = 0.0f / 0.0f;
  for (unsigned m = 0; m < g_metrics; m++) {
    float* column = slab + m * maxRows;
    for (unsigned r = 0; r < rows; r++) column[r] = nan;
  }
}

// Parse a block of WHOLE rows (`len` bytes from a row boundary, ending after
// a '\n'). `year` is the case's year from JS; every row is checked against it.
// Rows are emitted in byte order with rowHour[row] giving their hour. Source
// column `c` writes plane `c - KEY_COLS`; JS maps planes to the cube by
// trimmed header name.
__attribute__((export_name("parse_block")))
unsigned parse_block(unsigned len, unsigned year) {
  // Hoisted out of the per-field macro, where a global reload would be paid
  // millions of times.
  const unsigned maxRows = g_maxRows;
  const unsigned numMetrics = g_metrics;
  // Unconfigured: writing would go through address 0, which in wasm is real
  // memory and would corrupt the data segment silently.
  if (maxRows == 0u) return 0u;

  const unsigned char* b = inbuf;
  unsigned col = 0, fs = 0;
  unsigned rowDay = NO_DAY, rowYear = 0;
  // Slot this row will occupy, or NO_DAY while the row is being rejected.
  unsigned slot = NO_DAY;
  g_rows = 0; g_overflow = 0; g_wideField = 0; g_badRow = 0; g_feb29 = 0;
  g_yearMismatch = 0;

  const v128_t vcomma = wasm_i8x16_splat(',');
  const v128_t vnl    = wasm_i8x16_splat('\n');
  unsigned i = 0;

  #define FIELD(END)                                                            \
    {                                                                           \
      unsigned e = (END);                                                       \
      if (e > fs && b[e - 1] == '\r') e--;                                      \
      if (col == 0) {                                                           \
        rowDay = date_to_day(b + fs, b + e, &rowYear);                          \
      } else if (col == 1) {                                                    \
        unsigned hourOfDay = parse_uint(b + fs, b + e);                         \
        slot = NO_DAY;                                                          \
        if (rowDay == FEB29) {                                                  \
          g_feb29++;                     /* dropped on purpose */          \
        } else if (rowDay == NO_DAY || hourOfDay < 1 || hourOfDay > 24) {       \
          g_badRow++;                    /* unreadable Date or Hour */          \
        } else if (rowYear != year) {                                           \
          g_yearMismatch++;              /* a second year in one case */        \
        } else if (g_rows >= maxRows) {                                         \
          g_overflow++;                  /* more rows than the layout holds */  \
        } else {                                                                \
          slot = g_rows++;                                                      \
          rowHour[slot] = (unsigned short)(rowDay * 24u + (hourOfDay - 1u));    \
          rowTou[slot] = 0;                                                     \
        }                                                                       \
      } else if (col == 2) {                                                    \
        /* "OnPeak" / "OffPeak" differ at byte 1: 'n' vs 'f'. */                \
        if (slot != NO_DAY) {                                                   \
          rowTou[slot] = (unsigned char)((e > fs + 1 && b[fs + 1] == 'n') ? 1 : 0); \
        }                                                                       \
      } else {                                                                  \
        unsigned m = col - (unsigned)KEY_COLS;                                  \
        if (m >= numMetrics) g_wideField++;                                     \
        else if (slot != NO_DAY) {                                              \
          slab[m * maxRows + slot] = parse_float(b + fs, b + e);                \
        }                                                                       \
      }                                                                         \
    }

  for (; i + 16 <= len; i += 16) {
    v128_t chunk = wasm_v128_load(b + i);
    v128_t hit   = wasm_v128_or(wasm_i8x16_eq(chunk, vcomma), wasm_i8x16_eq(chunk, vnl));
    unsigned mask = (unsigned)wasm_i8x16_bitmask(hit);
    while (mask) {
      unsigned pos = i + (unsigned)__builtin_ctz(mask);
      mask &= mask - 1;
      FIELD(pos)
      fs = pos + 1;
      col++;
      /* g_rows counts EMITTED rows and is advanced when a slot is claimed, so
         a newline only resets the per-row state. */
      if (b[pos] == '\n') { col = 0; rowDay = NO_DAY; slot = NO_DAY; }
    }
  }
  for (; i < len; i++) {
    unsigned char c = b[i];
    if (c == ',' || c == '\n') {
      FIELD(i)
      fs = i + 1;
      col++;
      if (c == '\n') { col = 0; rowDay = NO_DAY; slot = NO_DAY; }
    }
  }
  #undef FIELD
  return g_rows;
}
